### PR TITLE
Draw over ItemDecoration and handle different row heights

### PR DESCRIPTION
**Pull request by tinsukE**
Some changes I had to make in order to use the library in a project.

### CODE REVIEW

File: lib-fastscrollrecyclerview/src/main/java/com/jaredrummler/fastscrollrecyclerview/FastScrollRecyclerView.java [hunk 1]
```diff
@@ -387,8 +390,58 @@ protected void getCurScrollState(ScrollPositionState stateOut) {
...
+        final int height = getHeight();
+        final int paddingTop = getPaddingTop();
+        final int paddingBottom = getPaddingBottom();
+
+        // How many rows are visible, like 10.5f for 10 rows completelly and one halfway visible
+        float visibleRows = 0f;
+
+        for (int position = firstVisiblePosition; position <= lastVisiblePosition; position++) {
+          final View itemView = findViewHolderForLayoutPosition(position).itemView;
+          final int itemHeight = itemView.getHeight();
```
Comment (jaredrummler) [inline lib-fastscrollrecyclerview/src/main/java/com/jaredrummler/fastscrollrecyclerview/FastScrollRecyclerView.java:423]: Possible NPE

File: lib-fastscrollrecyclerview/src/main/java/com/jaredrummler/fastscrollrecyclerview/FastScrollRecyclerView.java [hunk 2]
```diff
@@ -387,8 +390,58 @@ protected void getCurScrollState(ScrollPositionState stateOut) {
...
+
+          // Finds how much of the itemView is actually visible, this allows smooth changes of the scrollbar thumb height
+          final int visibleHeight = itemHeight
+                  - Math.max(0, paddingBottom - layoutManager.getDecoratedTop(itemView)) // How much is cut at the top
+                  - Math.max(0, paddingBottom + layoutManager.getDecoratedBottom(itemView) - height); // How much is cut at the bottom
+
+          visibleRows += visibleHeight / (float) itemHeight;
+        }
+
+        System.out.println("ROW HEIGHT: " + Math.round((height - (paddingTop + paddingBottom)) / visibleRows));
```
Comment (jaredrummler) [inline lib-fastscrollrecyclerview/src/main/java/com/jaredrummler/fastscrollrecyclerview/FastScrollRecyclerView.java]: Use android.util.LOG or remove this

File: lib-fastscrollrecyclerview/src/main/java/com/jaredrummler/fastscrollrecyclerview/FastScrollRecyclerView.java [hunk 3]
```diff
@@ -387,8 +390,58 @@ protected void getCurScrollState(ScrollPositionState stateOut) {
...
+  }
+
+  /**
+   * Calculates the row height based on the average of the visible children, to handle scrolling
+   * through children with different heights gracefully
+   */
+  protected int calculateRowHeight(int fallbackHeight) {
+    LayoutManager layoutManager = getLayoutManager();
+
+    if (layoutManager instanceof LinearLayoutManager || layoutManager instanceof GridLayoutManager) {
```
Comment (jaredrummler) [inline lib-fastscrollrecyclerview/src/main/java/com/jaredrummler/fastscrollrecyclerview/FastScrollRecyclerView.java]: GridLayoutManager extends from LinearLayoutManager so the second condition is always true.

**Comment by tinsukE** [comment]
Thanks for the review! I really overlooked some points. Please check the updates.

**Pull request merged**